### ISSUE TITLE
is the `have_lief` code path for the patchelf code broken here?

### ISSUE THREAD

**Issue by leontristain** [conda/conda-build#3588]
I'm getting a whole bunch of errors that look like this when trying to build a package:

```ERROR :: get_rpaths_raw()=['/home/buildbot/miniconda2/envs/buildenv/conda-bld/aafsdk_1562028238846/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib'] and patchelf=/home/buildbot/miniconda2/envs/buildenv/conda-bld/aafsdk_1562028238846/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib disagree for /home/buildbot/miniconda2/envs/buildenv/conda-bld/aafsdk_1562028238846/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib/libaafpgapi.so ::
ERROR :: get_rpaths_raw()=[] and patchelf= disagree for /home/buildbot/miniconda2/envs/buildenv/conda-bld/aafsdk_1562028238846/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib/libDNxHDCodec.so ::
ERROR :: get_rpaths_raw()=['/home/buildbot/miniconda2/envs/buildenv/conda-bld/aafsdk_1562028238846/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib'] and patchelf=/home/buildbot/miniconda2/envs/buildenv/conda-bld/aafsdk_1562028238846/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib disagree for /home/buildbot/miniconda2/envs/buildenv/conda-bld/aafsdk_1562028238846/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib/libaafintp.so ::
ERROR :: get_rpaths_raw()=['/home/buildbot/miniconda2/envs/buildenv/conda-bld/aafsdk_1562028238846/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib'] and patchelf=/home/buildbot/miniconda2/envs/buildenv/conda-bld/aafsdk_1562028238846/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib disagree for /home/buildbot/miniconda2/envs/buildenv/conda-bld/aafsdk_1562028238846/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib/libcom-api.so ::
```

Looking at the error text, it's throwing the error because `get_rpaths_raw()` is a list of string while `patchelf` is a string. Digging further into the code I found this snippet in `post.py`:

```python
    patchelf = external.find_executable('patchelf', prefix)
    try:
        existing = check_output([patchelf, '--print-rpath', elf]).decode('utf-8').splitlines()[0]
    except CalledProcessError:
        print('patchelf: --print-rpath failed for %s\n' % (elf))
        return
    if have_lief:
        existing2, _, _ = get_rpaths_raw(elf)
        if existing != existing2:
            print('ERROR :: get_rpaths_raw()={} and patchelf={} disagree for {} :: '.format(
                existing2, existing, elf))
```

... where `existing` comes from a `check_output` call's result decoded, splitlined, and taken the first part, so it's definitely a string. Whereas `existing2` is the first of the things returned by `get_rpaths_raw`, which comes from `os_utils/liefldd.py`:

```python
if have_lief:
    def get_rpathy_thing_raw_partial(file, elf_attribute, elf_dyn_tag):
        '''
        By raw we mean that no processing is done on them whatsoever. The values are taken directly from
        LIEF. For anything but Linux, this means an empty list.
        '''

        binary_format = None
        binary_type = None
        binary = ensure_binary(file)
        rpaths = []
        if binary:
            binary_format = binary.format
            if binary_format == lief.EXE_FORMATS.ELF:
                binary_type = binary.type
                if binary_type == lief.ELF.ELF_CLASS.CLASS32 or binary_type == lief.ELF.ELF_CLASS.CLASS64:
                    rpaths = _get_elf_rpathy_thing(binary, elf_attribute, elf_dyn_tag)
            elif (binary_format == lief.EXE_FORMATS.MACHO and
                  binary.has_rpath and
                  elf_dyn_tag == lief.ELF.DYNAMIC_TAGS.RPATH):
                rpaths.extend([command.path for command in binary.commands
                  if command.command == lief.MachO.LOAD_COMMAND_TYPES.RPATH])
        return rpaths, binary_format, binary_type
    get_runpaths_raw = partial(get_rpathy_thing_raw_partial, elf_attribute='runpath', elf_dyn_tag=lief.ELF.DYNAMIC_TAGS.RUNPATH)
    get_rpaths_raw = partial(get_rpathy_thing_raw_partial, elf_attribute='rpath', elf_dyn_tag=lief.ELF.DYNAMIC_TAGS.RPATH)
```

... and that's definitely a list. So then the two would never be the same and you'd always get the error.

Unless I'm reading the code wrong, this looks like a bug, hence reporting it.

Thanks!

**Comment by leontristain** [comment]
Actually, looking at the code again, the 'ERROR' text is merely printed and doesn't cause any exceptions to be raised or the function to early exit in some other way, so maybe this is actually harmless and only "looks" like a problem? Seems like I might have gotten distracted. If so, apologies.

**Comment by mingwandroid** [comment]
Maybe, maybe not, the code is being rewritten though so I'd not spend too much time on it.

**Comment by cgr71ii** [comment]
I have been facing this problem as well. It is true that the problem does not stop the execution of the build, but is really annoying (I am building a project where are lots of binaries, and the quantity of warnings due to this issue are enough to skip all them and hope the best because you do not which of them might be a truly problem). The cause of this comment is to inform to all those which, as me, reach this issue and let them know that this problem has been fixed (do not know if totally or just partially) in commit  4e94370.

Currently, the last available release is 3.25.5, but i suppose that this fix will be released in next.

**Comment by mbargull** [comment]
@cgr71ii, ah, I didn't find this issue before! Yes, gh-4131 should've resolved this and will be in `conda-build >=3.20.6`.
